Add MovementAction to Cpt. Toad

andreafra opened the pull request:
Cpt. Toad moves relatively to the camera position and angle

Also removed a couple of debug leftovers

andreafra committed:
Move Cpt. Toad relative to Camera POV

Also remove a couple of debug leftovers

QUB3X commented:
It'd be nice to make the Toad rotate using the `alpha` angle in his Movement action.

aleasto commented on src/engine/Core.ts:
```diff
@@ -4,6 +4,7 @@ import { Light } from "./Lights";
 import { Node, RenderNode, State } from "./SceneGraph";
 import * as DebugLine from "./debug/Lines";
 import { WebGLProgramInfo } from "./Shaders";
+import { GetMode } from "../main";
```
No

aleasto commented on src/engine/Core.ts:
```diff
@@ -51,7 +52,7 @@ function Render(time: DOMHighResTimeStamp) {
 	}
 
-	DebugLine.Render(viewProjectionMatrix);
+	/* if (GetMode() === "EDITOR") */ DebugLine.Render(viewProjectionMatrix);
```
No

aleasto commented on src/engine/SceneGraph.ts:
```diff
@@ -4,6 +4,7 @@ import * as Engine from "./Core";
 import { Light } from "./Lights";
 import { WebGLProgramInfo } from "./Shaders";
 import { gl } from "./Core";
+import { GetMode } from "../main";
```
No

aleasto commented on src/engine/SceneGraph.ts:
```diff
@@ -89,13 +90,14 @@ export class Node<T extends State> {
 		this.UpdateWorldMatrix(worldMatrix);
-		this.ExecuteActions();
+
+		if (GetMode() === "GAME") this.ExecuteActions(deltaTime);
```
No

aleasto commented on src/models/Toad.ts:
```diff
@@ -104,9 +111,41 @@ export function Spawn() {
 	return toadNode;
 }
+
+const MovementAction = (deltaTime: number, state: ToadState): void => {
```
if (GetMode() != "GAME") return;

aleasto commented on src/models/Toad.ts:
```diff
@@ -104,9 +111,41 @@ export function Spawn() {
 }
+
+const MovementAction = (deltaTime: number, state: ToadState): void => {
+	let position = utils.ComputePosition(state.localMatrix, [0, 0, 0]);
```
worldMatrix?

aleasto committed:
Compute the ViewProjectionMatrix after all nodes were updated

Some nodes may update the camera or projection, so render them
with the updated matrix rather than having to wait one frame

aleasto committed:
Remove engine dependencies on main

aleasto committed:
Make toad rotate with the camera

This is a big hack that would ignore any scaling or shear that the
Toad node might be spawned with

aleasto committed:
Make toad look in their movement direction instead
